vopono_core/src/network/piapf.rs

```diff
@@ -0,0 +1,193 @@
+extern crate json;
+
+use std::sync::mpsc::{self, Receiver};
+use std::{
+    sync::mpsc::Sender,
+    thread::JoinHandle
+};
+use base64::prelude::*;
+use regex::Regex;
+
+use super::netns::NetworkNamespace;
+use super::Forwarder;
+
+use crate::config::vpn::Protocol;
+use crate::config::providers::OpenVpnProvider; // Added load_openvpn_auth to this trait 
+use crate::config::providers::pia::PrivateInternetAccess;  // Added load_wireguard_auth to this struct 
+
+/// Used to provide port forwarding for ProtonVPN
+pub struct Piapf {
+    pub port: u16,
+    loop_thread_handle: Option<JoinHandle<()>>,
+    send_channel: Sender<bool>,
+}
+
+struct ThreadParams {
+    pub port: u16,
+    pub netns_name: String,
+    pub signature: String, 
+    pub payload: String,
+    pub hostname: String,
+    pub gateway: String,
+}
+
+impl Piapf {
+    pub fn new(ns: &NetworkNamespace, protocol: &Protocol) -> anyhow::Result<Self> {
+        let pia = PrivateInternetAccess {}; //This is a bit weird, no? There's no state, so effectively all the methods are static...
+        
+        let traceroute_response = NetworkNamespace::exec_with_output(
+            &ns.name, 
+            &["traceroute", "-n", "-m", "1", "privateinternetaccess.com" ], )?;
```

Comment:
We should check `traceroute` is installed with the `which` crate and prompt to install it otherwise.